parquet-hadoop/src/main/java/org/apache/parquet/hadoop/RowGroupMerger.java

```diff
@@ -0,0 +1,634 @@
+
+            try {
+              reader.initFromPage(page.getValueCount(), page.getData().toInputStream());
+            } catch (IOException e) {
+              throw new UncheckedIOException(e);
+            }
+
+            for (int i = 0; i < page.getValueCount() - page.getNullCount(); i++) {
+              dataWriter.accept(reader, newValuesWriter);
+            }
+
+            BytesInput dataBytes = newValuesWriter.getBytes();
+            Encoding dataEncoding = newValuesWriter.getEncoding();
+            int uncompressedSize = Math.toIntExact(dataBytes.size());
+            BytesInput compressedBytes = compress(dataBytes);
+
+            newValuesWriter.reset();
+
+            return DataPageV2.compressed(
+              page.getRowCount(), page.getNullCount(), page.getValueCount(),
+              page.getRepetitionLevels(), page.getDefinitionLevels(), dataEncoding,
+              compressedBytes, uncompressedSize, page.getStatistics()
+            );
+
+          } else { //not dictionary encoded
+
+            return DataPageV2.compressed(
+              page.getRowCount(), page.getNullCount(), page.getValueCount(),
+              page.getRepetitionLevels(), page.getDefinitionLevels(), page.getDataEncoding(),
+              compress(page.getData()), page.getUncompressedSize(), page.getStatistics());
+          }
+        }
+      });
+    }
+
+    private void verifyDataEncoding(Encoding encoding, Dictionary pageDictionary) {
+      if (encoding.usesDictionary() && pageDictionary == null) {
+        throw new ParquetDecodingException(
+          format("could not read page in column %s as the dictionary was missing for encoding %s", columnDesc, encoding));
+      }
+    }
+
+    //expensive should be used as a last resort
+    private int calculateRowCount(DataPageV1 pageV1, ByteBufferInputStream in) {
+      try {
+        ValuesReader rlReader = pageV1.getRlEncoding().getValuesReader(columnDesc, REPETITION_LEVEL);
+        rlReader.initFromPage(pageV1.getValueCount(), in);
+        int rowCount = 0;
+        for (int i = 0; i < pageV1.getValueCount(); i++) {
+          rowCount = rlReader.readInteger() == 0 ? rowCount + 1 : rowCount;
+        }
+        return rowCount;
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+  }
+
+  private static class SizeEstimator {
+    private final boolean targetCompressed;//is destination block compressed
+    private double factor;
+
+    private SizeEstimator(boolean targetCompressed) {
+      this.targetCompressed = targetCompressed;
+    }
+
+    private long estimate(BlockMetaData blockMeta) {
+      if (factor <= 0) {
+        return targetCompressed ? blockMeta.getCompressedSize() : blockMeta.getTotalByteSize();//very naive estimator
+      }
+      return (long) (factor * blockMeta.getTotalByteSize());
+    }
+
+    private void update(long actualBytes, BlockMetaData blockMeta) {
```

Comment:
'factor' should be mandatorily initialized before it can be used. Would it be better to move it to the constructor?

Reply:
It is updated on every iteration so that the estimation can use the most recent calculated value. On creation of the estimator we do not have any info to calculate the factor